### PR TITLE
Optimizer: OpModuleProcessed is in its own layout section

### PR DESCRIPTION
**Pull request by dneto0**
This is a recent decision from the SPIR WG.  The spec update
has not yet been published.
Khronos SPIR-V internal issue 199

### CODE REVIEW

**Comment by dneto0** [comment]
FYI.  The validator already knows about this change about where OpModuleProcessed can go.
See c843ef8ab

File: source/opt/reflect.h [hunk 1]
```diff
@@ -26,12 +26,14 @@ namespace ir {
 
 inline bool IsDebug1Inst(SpvOp opcode) {
   return (opcode >= SpvOpSourceContinued && opcode <= SpvOpSourceExtension) ||
-         opcode == SpvOpString || opcode == SpvOpLine ||
-         opcode == SpvOpNoLine || opcode == SpvOpModuleProcessed;
+         opcode == SpvOpString || opcode == SpvOpLine || opcode == SpvOpNoLine;
```
Comment (antiagainst) [inline source/opt/reflect.h]: We cannot have `OpLine` and `OpNoLine` until the type-consant-variable section (9). So So I assume debug1 instructions should not contain them?

File: source/opt/reflect.h [hunk 2]
```diff
@@ -26,12 +26,14 @@ namespace ir {
...
 }
 inline bool IsDebug2Inst(SpvOp opcode) {
   return opcode == SpvOpName || opcode == SpvOpMemberName ||
-         opcode == SpvOpNoLine || opcode == SpvOpModuleProcessed;
+         opcode == SpvOpNoLine;
```
Comment (antiagainst) [inline source/opt/reflect.h]: Should remove `OpNoLine`.

File: source/opt/module.h [hunk 3]
```diff
@@ -74,6 +74,9 @@ class Module {
   // "debug 2" instructions are the ones in layout section 7.b), see section
   // 2.4 Logical Layout of a Module from the SPIR-V specification.
   inline void AddDebug2Inst(std::unique_ptr<Instruction> d);
+  // Appends a debug 3 instruction (OpModuleProcessed) to this module.
```
Comment (antiagainst) [inline source/opt/module.h]: Comments for debug1 and debug2 should be updated.

File: source/opt/module.h [hunk 4]
```diff
@@ -118,20 +121,35 @@ class Module {
...
 
   // Iterators for debug 2 instructions (excluding OpLine & OpNoLine) contained
-  // in this module.
+  // in this module.  These are for layout section 7b.
   inline inst_iterator debug2_begin();
   inline inst_iterator debug2_end();
   inline IteratorRange<inst_iterator> debugs2();
   inline IteratorRange<const_inst_iterator> debugs2() const;
 
+  // Iterators for debug 3 instructions (excluding OpLine & OpNoLine) contained
```
Comment (antiagainst) [inline source/opt/module.h]: Comments for debug1, debug2, & debug3 should be updated.

File: source/opt/module.h [hunk 5]
```diff
@@ -143,14 +161,17 @@ class Module {
...
-  void debug_clear() { debug1_clear(); debug2_clear(); }
+  void debug_clear() { debug1_clear(); debug2_clear(); debug3_clear(); }
 
   // Clears all debug 1 instructions (excluding OpLine & OpNoLine).
   void debug1_clear() { debugs1_.clear(); }
 
   // Clears all debug 2 instructions (excluding OpLine & OpNoLine).
   void debug2_clear() { debugs2_.clear(); }
 
+  // Clears all debug 3 instructions (excluding OpLine & OpNoLine).
```
Comment (antiagainst) [inline source/opt/module.h]: Comments regarding debug instructions need to be updated.

**Comment by dneto0** [comment]
Rebased

**Pull request merged**